LinuxAdapter.cs

```diff
@@ -0,0 +1,120 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using System;
+using System.DirectoryServices.Protocols;
+using System.Linq;
+using System.Net;
+using System.Security.Claims;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Authentication.Negotiate
+{
+    public class LinuxAdapter
+    {
+        private readonly string _distinguishedName;
+        private readonly LdapConnection _connection;
+
+        public LinuxAdapter(string machineAccount, string machinePassword)
+        {
+            if (string.IsNullOrEmpty(machineAccount))
+            {
+                throw new ArgumentNullException(nameof(machineAccount));
+            }
+            if (string.IsNullOrEmpty(machinePassword))
+            {
+                throw new ArgumentNullException(nameof(machinePassword));
+            }
+
+            var domain = machineAccount.Substring(machineAccount.IndexOf('@') + 1);
+            _distinguishedName = domain.Split('.').Select(name => $"dc={name}").Aggregate((a, b) => $"{a},{b}");
+
+            LdapDirectoryIdentifier di = new LdapDirectoryIdentifier(server: domain, fullyQualifiedDnsHostName: true, connectionless: false);
+            NetworkCredential credential = new NetworkCredential(machineAccount, machinePassword);
+            _connection = new LdapConnection(di, credential);
+            _connection.SessionOptions.ProtocolVersion = 3; //Setting LDAP Protocol to latest version
+            _connection.Bind(); // This line actually makes the connection.
```

Comment:
Oh dear. Bind isn't a huge concern since it only happens once, but SendRequest MUST be async or it will block the thread pool and take down servers. What's the plan for fixing async?

Reply:
One other question for @joperezr - does the library support secure LDAP, i.e. LDAPS? Windows will default to that soon.

Reply:
There is a proposal to add SendRequest versions which are async https://github.com/dotnet/runtime/issues/37415 and there is also another issue to fix BeginSendRequest https://github.com/dotnet/runtime/issues/36690 but none of them are actively been worked on that I’m aware of.

Regarding LDAPS, I believe it is supported as there are ways to use SSL to bind against an LDAP server using openldap but it probably requires a bit of work to get it working.

Reply:
@blowdart, @joperezr, we'll need one of those two addressed for 5.0. Sync IO in a web server is not an option.

Reply:
LDAPS also needs to be looked at, before AD switches to it completely.

Reply:
>we'll need one of those two addressed for 5.0. Sync IO in a web server is not an option.

Adding SendRequestAsync methods will take longer as it will require going through API Review as well as a full implementation, whereas for fixing BeginSendRequest should be much easier and faster as all we need to do there is to switch from APM async model (not supported in .NET Core) to Task-based so I can do that today.

Reply:
@Tratcher I have sent out https://github.com/dotnet/runtime/pull/37774 which will fix LdapConnection.BeginSendRequest and LdapConnection.EndSendRequest so after that is merged you can start depending on that instead in order to not block a thread in the Threadpool. Unfortunatelly the native libraries we depend on don't really support callback-based async for sending requests, but they do support polling to check if the results are ready so that is what I used to implement it.

Reply:
... at least you're polling async.